hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioPipeline.java

```diff
@@ -385,4 +386,71 @@ public void run() {
             this.next = next;
         }
     }
+
+    interface PublishableMultiCounter<K> {
+        void inc(K key, long amount);
+        long publish(K key);
+    }
+
+    static class NoOpCounter<K> implements PublishableMultiCounter<K> {
+        @Override
+        public void inc(K key, long amount) {
+            // NO-OP
+        }
+
+        @Override
+        public long publish(K key) {
+            return 0;
+        }
+    }
+
+    static class PublishableMultiCounterImpl<K extends Enum<K>> implements PublishableMultiCounter<K> {
+        private final EnumMap<K, PublishableSwCounter> counters;
+
+        PublishableMultiCounterImpl(Class<K> keyType) {
+            counters = new EnumMap<K, PublishableSwCounter>(keyType);
+            for (K key : keyType.getEnumConstants()) {
+                counters.put(key, new PublishableSwCounter());
+            }
+        }
+
+        @Override
+        public void inc(K key, long amount) {
+            counters.get(key).inc(amount);
+        }
+
+        /**
+         * Publishes the bytes transceived since the last publish.
+         *
+         * @return number of bytes transceived since this method was last called
+         */
+        @Override
+        public long publish(K key) {
+            return counters.get(key).publishDelta();
+        }
+
+        /**
+         * A single writer counter that remembers its last published
+         * value and can publish the delta since the last publish.
+         */
+        private static final class PublishableSwCounter {
+            private final SwCounter counter = newSwCounter();
+            private long lastPublishedValue;
```

Comment:
Looks like this field can be accessed from multiple threads, so it's worth converting it to something more thread-safe, like `volatile`. That's because `NioNetworking.NioThread` is executed on `MetricsRegistryImpl#scheduler` which has the size of 2 threads. WDYT?

Reply:
In my experiments, it was always the same thread updating the value, but you're right - there's no guarantee to it AFAICT. It should be volatile. Let me update.